heron/api/src/java/com/twitter/heron/api/Config.java

```diff
@@ -100,13 +100,65 @@
    */
   public static final String TOPOLOGY_SERIALIZER_CLASSNAME = "topology.serializer.classname";
   /**
-   * How many executors to spawn for ackers.
+   * Is the topology running in atleast-once mode?
    * <p>
-   * <p>If this is set to 0, then Heron will immediately ack tuples as soon
+   * <p>If this is set to false, then Heron will immediately ack tuples as soon
    * as they come off the spout, effectively disabling reliability.</p>
+   * This is deprecated. Please use TOPOLOGY_RELIABILITY_MODE instead.
+   * This config parameter will be removed in future releases
    */
   public static final String TOPOLOGY_ENABLE_ACKING = "topology.acking";
   /**
+   * What is the reliability mode under which we are running this topology
+   * Topology writers must set TOPOLOGY_RELIABILITY_MODE to one
+   * one of the following modes
+   */
+  public enum TopologyReliabilityMode {
+    /**
+     * Heron provides no guarantees wrt tuple delivery. Tuples emitted by
+     * components can get lost for any reason(network issues, component failures,
+     * overloaded downstream component, etc).
+     */
+    ATMOST_ONCE,
+    /**
+     * Heron guarantees that each emitted tuple is seen by the downstream components
+     * atleast once. This is achieved via the anchoring process where emitted tuples
+     * are anchored based on input tuples. Note that in failure scenarios, downstream
+     * components can see the same tuple multiple times.
+     */
+    ATLEAST_ONCE,
+    /**
+     * In this mode Heron will try to take the snapshots of
+     * all of the components of the topology every
+     * TOPOLOGY_STATEFUL_CHECKPOINT_INTERVAL_SECONDS seconds. Upon failure of
+     * any component, its state is recovered to the last checkpoint saved. The
+     * spouts/bolts of the topology have to implement IStatefulComponent for state
+     * to be actually captured.
+     * While this mode provides a way for topology components to recover their state
+     * upon their death, it does not guard against tuple losses that can happen because
+     * of network partitions and other reasons.
+     */
+    STATEFUL,
```

Comment:
This seems orthogonal to the reliability mode IMO. If exactly once leverages the stateful impl that's an implementation detail. From the users perspective these are different things. The reliability mode should specify guarantees with tuple reliability, while the stateful mode should define whether herons state management is being leveraged and enabled by the user logic in the topology.

IIRC the reason for requiring the stateful enabled flag was so that users could turn stateful processing of their topology on or off. If that's the case they might choose to do either while still keeping exactly once enabled.